### PR TITLE
Start to add lots of documentation

### PR DESCRIPTION
**Pull request by carols10cents**
I started this a while ago, but an email from @vignesh-sankaran prompted me to get what I have out there rather than bitrotting on my machine.

I'd love a review of what I have so far, @vignesh-sankaran! What do you think of the outline of the architecture document? What kinds of things would make that document a more useful high-level overview? 

The doc comments should hopefully make the docs at https://docs.rs/cargo-registry/ better, I don't think the little bit I've added here warrants a version bump and publish yet, though.

### CODE REVIEW

**Comment by jtgeibel** [comment]
Hey @carols10cents, here are my cleaned up notes from going through the layout of my working directory.  I have this as `docs/REPO_LAYOUT.md` for now.

# Repo File Layout

## Backend

* `build.rs` - Cargo build script
* `Cargo.lock` - Locks dependencies to specific versions providing consistency across development and deployment
* `Cargo.toml` - Defines the crate and its dependencies
* `migrations/` - Diesel migrations applied to the database during development and deployment
* `.rustfmt.toml` - Defines Rust coding style guidelines which are enforced by the CI environment
* `src/` - The backend's source code
* `target/` - Compiled output, including dependencies and final binary artifacts - (ignored in `.gitignore`)
* `tmp/index-co` - The registry repository; in production this is cloned from Github and in development from `tmp/index-bare` - (ignored in `.gitignore`)

## Front End

* `app/` - The frontend's source code
* `config/{environment,targets}.js` - Configuration of the frontend
* `dist/` - Contains the distributable (optimized and self-contained) output of building the frontend; served under the root `/` url - (ignored in `.gitignore`)
* `.ember-cli` - Settings for the `ember` command line interface
* `ember-cli-build.js` - Contains the build specification for Broccoli
* `.eslintrc.js` - Defines Javascript coding style guidelines (enforced during CI???)
* `mirage/` - A mock backend used during development and testing
* `node_modules/` - npm dependencies - (ignored in `.gitignore`)
* `package.json` - Defines the npm package and its dependencies
* `package-lock.json` - Locks dependencies to specific versions providing consistency across development and deployment
* `public/` - Static files that are merged into `dist/` during build
* `testem.js` - Integration with Test'em Scripts
* `tests/` - Frontend tests
* `vendor/` - frontend dependencies not distributed by npm; not currently used

## Deployment - Heroku

* `app.json` - Configuration for Heroku Deploy Button
* `.buildpacks` - A list of buildpacks used during deployment
* `config/nginx.conf.erb` - Template used by the nginx buildpack
* `.diesel_version` - Used by diesel buildpack to install a specific version of Diesel CLI during deployment
* `Procfile` - Contains process type declarations for Heroku

## Development

* `docs/` - The location of design documentation
* `.editorconfig` - Coding style definitions supported by some IDEs // TODO: Reference extensions for common editors
* `.env` - Environment variables loaded by the backend - (ignored in `.gitignore`)
* `.env.sample` - Example environment file checked into the repository
* `.git/` - The git repository; not available in all deployments (e.g. Heroku)
* `.gitignore` - Configures git to ignore certain files and folders
* `script/init-local-index.sh` - Creates registry repositories used during development
* `tmp/` - Temporary files created during development; when deployed on Heroku this is the only writable directory - (ignored in `.gitignore`)
* `tmp/index-bare` - A bare git repository, used as the origin for `tmp/index-co` during development - (ignored in `.gitignore`)
* `.travis.yml` - Configuration for continous integration at https://travis-ci.org/rust-lang/crates.io
* `.watchmanconfig` - Use by Ember CLI to efficiently watch for file changes if you install watchman

## Other

* `LICENSE-APACHE`
* `LICENSE-MIT`
* `README.md`

File: docs/ARCHITECTURE.md [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+
+#### The `http` module
+
+#### The `model` module
+
+#### The `schema` module
+
+#### The `utils` module
+
+### Code having to do with managing a registry of crates
```
Comment (vignesh-sankaran) [inline docs/ARCHITECTURE.md]: Hmm, not quite sure where you're referring to with this?

**Comment by vignesh-sankaran** [comment]
Thanks for starting on this @carols10cents, the structure of the document itself looks good, though I think we should mention that it's specifically for the backend. Maybe we could look into having a similar document for the front end too?

**Comment by carols10cents** [comment]
> though I think we should mention that it's specifically for the backend. Maybe we could look into 
> having a similar document for the front end too?

Yeah, I have [a heading in there for backend](https://github.com/rust-lang/crates.io/pull/899/files#diff-264f0dd1ee08f0efe10ba3bcc1b93eebR9) and I think I just got tired before making a corresponding frontend heading :)

> Hmm... could we look at removing "The" in front of the headers in docs/ARCHITECTURE.md that refer to backend modules? There's a lot of them :)

🤷‍♀️ That's how I'd talk about them though... "Where's the code to do x?" "it's in the `db` module" 

> Also, maybe we could look at adding a diagram of the database schema?

I'd be into that, if there's a tool we can use to automate creating the diagram and if we keep it up to date :) Do you know of a good one?

File: docs/ARCHITECTURE.md (hunk 1 above)
Comment (carols10cents) [inline docs/ARCHITECTURE.md]: I'm not sure why you're not sure :) What are the possible interpretations that you're deciding between? Note that this has an `h3`, and the `h4`s from line 102 to 126 are under this heading, does that help? What would be a better heading?

**Comment by carols10cents** [comment]
Ok -- I've incorporated @jtgeibel's repo layout docs and took @vignesh-sankaran's suggestion to break this out into separate backend and frontend documents-- I could see this getting long :)

WDYT?

File: docs/ARCHITECTURE.md [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+## Backend - Rust
+
+The backend of crates.io is written in Rust. Most of that code lives in the *src* directory. It
+serves a JSON API over HTTP, and the HTTP server interface is provided by the [conduit][] crate and
+related crates. More information about the backend is in
+[`docs/BACKEND.md`](https://github.com/rust-lang/crates.io/blob/master/docs/BACKEND.md).
+
+[conduit]: https://crates.io/crates/conduit
+
+These files have to do with the backend:
```
Comment (vignesh-sankaran) [inline docs/ARCHITECTURE.md]: How about "these files and directories"?

File: docs/ARCHITECTURE.md (hunk 1 above)
Comment (vignesh-sankaran) [inline docs/ARCHITECTURE.md]: I was trying to figure out how this fit into one module, yes I didn't see the `h3` for it compared to the `h4`'s that refer to the modules

File: docs/ARCHITECTURE.md [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+hopefully after reading this doc, you'll have a good idea of where to start looking for the code
+you want to change.
+
+This is a work in progress. Pull requests and issues to improve this document are very welcome!
+
+## Documentation
+
+Documentation about the codebase appears in these locations:
+
+* `LICENSE-APACHE` and `LICENSE-MIT` - the terms under which this codebase is licensed.
```
Comment (jtgeibel) [inline docs/ARCHITECTURE.md]: It looks like `package.json` only lists MIT.  I'm not sure if that is an oversight or if we need to word this more carefully.  We should probably explicitly call out dual licensing in `README.md` 

Also see: https://docs.npmjs.com/files/package.json#license

**Comment by carols10cents** [comment]
I've addressed the two things yinz pointed out, and in the interest of getting this merged so that we can all start working on this, and because this is all just docs, I'm going to merge my own PR 😎 

bors r+

**Pull request merged**